### PR TITLE
feat: use shallow clones for backups

### LINKED ISSUE #209: Shallow clone Git repository during backup
Right now, we clone the whole working tree, which slows down the backup. It is sufficient to do a shallow clone.

For existing branches, shallow clone the branch. For non-existing branches, shallow clone the default branch and create a new one.

### PR DESCRIPTION
**Pull request by amolenaar**
Checkout only the latest commit for a branch, just enough to create a new commit upon.

Fixes #209.

### CODE REVIEW

File: t4c/backup.py [hunk 1]
```diff
@@ -140,28 +140,40 @@ def clone_git_repository() -> pathlib.Path:
...
             check=True,
             cwd=git_dir,
+            env=env,
         )
     except subprocess.CalledProcessError as e:
         print(e.returncode)
         if e.returncode == 128:
             subprocess.run(
-                ["git", "switch", "-c", os.environ["GIT_REPO_BRANCH"]],
+                [
```
Comment (MoritzWeber0) [inline t4c/backup.py:167]: Have you tested this case? Where do you create the new branch as described in the #209? (For non-existing branches, shallow clone the default branch and create a new one.)

I didn't start the local development environment with T4C server and Git server, but `git push origin branch2` should fail if there the branch `branch2` doesn't exist locally.

**Review by MoritzWeber0 — CHANGES_REQUESTED** [review]
Looks good, but haven't tested it.

File: t4c/backup.py (hunk 1 above)
Comment (amolenaar) [reply t4c/backup.py:167]: I tested it with Git locally. It does use the default branch defined on the remote.

**Comment by amolenaar** [comment]
Here's the output of some more thorough testing of the backup process.


## Preparation

Created a bare repository:

```
git init --bare myrepo
```

## First backup on fresh branch

```
[master (root-commit) 2369e91] Backup
 6 files changed, 833 insertions(+)
 create mode 100644 .project
 create mode 100644 CommitHistory.activitymetadata
 create mode 100644 CommitHistory.json
 create mode 100644 Test.afm
 create mode 100644 Test.aird
 create mode 100644 Test.capella
Enumerating objects: 8, done.
Counting objects: 100% (8/8), done.
Delta compression using up to 8 threads
Compressing objects: 100% (8/8), done.
Writing objects: 100% (8/8), 9.04 KiB | 9.04 MiB/s, done.
Total 8 (delta 1), reused 0 (delta 0), pack-reused 0
To /tmp/myrepo
 * [new branch]      HEAD -> my-branch
INFO:Importer:Backup of model finished
```

## Second backup on existin gbranch

```
[my-branch 4d5ab1f] Backup
 2 files changed, 176 insertions(+), 176 deletions(-)
Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 8 threads
Compressing objects: 100% (4/4), done.
Writing objects: 100% (4/4), 721 bytes | 721.00 KiB/s, done.
Total 4 (delta 3), reused 0 (delta 0), pack-reused 0
To /tmp/myrepo
   2369e91..4d5ab1f  HEAD -> my-branch
INFO:Importer:Backup of model finished
```

## Check the repo

```
root@fedora:~# git clone --branch=my-branch /tmp/myrepo
Cloning into 'myrepo'...
done.
root@fedora:~# cd myrepo/
root@fedora:~/myrepo# ls
CommitHistory.activitymetadata  CommitHistory.json  Test.afm  Test.aird  Test.capella
root@fedora:~/myrepo# git log
commit 4d5ab1fb9bc865d957ef853bc68d2c2775d06ffd (HEAD -> my-branch, origin/my-branch)
Author: user <backup@example.com>
Date:   Fri Oct 13 16:41:39 2023 +0000

    Backup

commit 2369e91ccfd5c388fa12c35c0d998c85ede0ea8f
Author: user <backup@example.com>
Date:   Fri Oct 13 16:39:02 2023 +0000

    Backup

root@fedora:~/myrepo#
```

**Review by MoritzWeber0 — APPROVED** [review]
Nice improvement and thanks for testing it 🥳

**Pull request merged**